tempesta_fw/tls.c

```diff
@@ -482,6 +482,13 @@ tfw_tls_encrypt(struct sock *sk, struct sk_buff *skb, unsigned int limit)
 	for (p = pages; p < pages_end; ++p)
 		put_page(*p);
 
+	if (type == TTLS_MSG_ALERT &&
+	    (io->alert[1] == TTLS_ALERT_MSG_CLOSE_NOTIFY ||
+	     io->alert[0] == TTLS_ALERT_LEVEL_FATAL))
+	{
+		ss_close(sk, SS_F_SYNC);
```

Comment:
We've discussed in chat about different problems related to reference counting, and I happy to see a solution without that caveats. At least it looks so for me now. Please check all the cases we have discussed in the chat, but I think we won't have any problems there.

An optimisation is possible here, but it's quite tricky. And can be done as enhancement later.  `ss_close()` adds a task into work_queue, calls `connection_drop()` hook and invokes `ss_do_close()` which will create a new skb with TCP_FIN flag. Instead, we can move some logic from `ss_do_close` right here, and append TCP_FIN to the alert directly.  Then we need to somehow invoke `connection_drop` or `connection_error` callback as `ss_close` does.

UPD: I see one possible issue. Imagine, there are a lot of requests in processing state, and corresponding responses are continuously added to the write_queue. Then an error happens on receiving path, and close notify is inserted between responses in write_queue. So we call `ss_close()` here, which schedules connection close. `TCP_FIN` flag will be added to the last skb in the queue - response after fatal alert. And we will send a little more data, than we wanted. Can't call it critical issue, though. With the enhancement above such befaviour can be coped.

Reply:
I agree with @ikoveshnikov and the solution is really elegant . It seems there are actually not so much code to write to mark the `skb` with `FIN` and move the socket to the closing state. Just please make it in a function in `sock.c`. It seems `skb && tcp_send_head(sk)` is always true in this function, but the rest of `ss_do_close()` should be the same, so probably `ss_do_close()` can be reworked to use the same subfunction for the new function plus the code with an skb allocation.